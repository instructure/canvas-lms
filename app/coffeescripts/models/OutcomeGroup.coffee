#
# Copyright (C) 2012 Instructure, Inc.
#
# This file is part of Canvas.
#
# Canvas is free software: you can redistribute it and/or modify it under
# the terms of the GNU Affero General Public License as published by the Free
# Software Foundation, version 3 of the License.
#
# Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
# A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
# details.
#
# You should have received a copy of the GNU Affero General Public License along
# with this program. If not, see <http://www.gnu.org/licenses/>.
#

define [
  'Backbone'
], (Backbone) ->

  class OutcomeGroup extends Backbone.Model

    name: ->
      @get 'title'

    # The api returns abbreviated data by default
    # which in most cases means there's no description.
    # Run fetch() to get all the data.
    isAbbreviated: ->
      !@has('description')

    setUrlTo: (action) ->
      @url =
        switch action
          when 'add' then @get('parent_outcome_group').subgroups_url
          when 'edit', 'delete' then @get 'url'
