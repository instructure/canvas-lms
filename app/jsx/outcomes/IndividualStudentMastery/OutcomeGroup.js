/*
 * Copyright (C) 2018 - present Instructure, Inc.
 *
 * This file is part of Canvas.
 *
 * Canvas is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3 of the License.
 *
 * Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import React from 'react'
import PropTypes from 'prop-types'
import ImmutablePropTypes from 'react-immutable-proptypes'
import I18n from 'i18n!outcomes'
import Flex, { FlexItem } from '@instructure/ui-layout/lib/components/Flex'
import View from '@instructure/ui-layout/lib/components/View'
import ToggleGroup from '@instructure/ui-toggle-details/lib/components/ToggleGroup'
import List, { ListItem } from '@instructure/ui-elements/lib/components/List'
import Pill from '@instructure/ui-elements/lib/components/Pill'
import Text from '@instructure/ui-elements/lib/components/Text'
import natcompare from 'compiled/util/natcompare'
import Outcome from './Outcome'
import * as shapes from './shapes'

const outcomeGroupHeader = (outcomeGroup, numMastered, numGroup) => (
  <Flex justifyItems="space-between">
    <FlexItem padding="0 x-small 0 0"><Text size="large" weight="bold">{ outcomeGroup.title }</Text></FlexItem>
    <FlexItem><Pill text={I18n.t('%{numMastered} of %{numGroup} Mastered', { numMastered, numGroup })} /></FlexItem>
  </Flex>
)

export default class OutcomeGroup extends React.Component {
  static propTypes = {
    outcomeGroup: shapes.outcomeGroupShape.isRequired,
    outcomes: PropTypes.arrayOf(shapes.outcomeShape).isRequired,
    expanded: PropTypes.bool.isRequired,
    expandedOutcomes: ImmutablePropTypes.set.isRequired,
    onExpansionChange: PropTypes.func.isRequired,
    outcomeProficiency: shapes.outcomeProficiencyShape
  }

  static defaultProps = {
    outcomeProficiency: null
  }

  handleToggle = (_event, expanded) => {
    this.props.onExpansionChange('group', this.props.outcomeGroup.id, expanded)
  }

  render () {
    const { outcomeGroup, outcomes, expanded, expandedOutcomes, onExpansionChange, outcomeProficiency } = this.props
    const numMastered = outcomes.filter((o) => o.mastered).length
    const numGroup = outcomes.length

    return (
      <View as="div" className="outcomeGroup">
        <ToggleGroup
          summary={outcomeGroupHeader(outcomeGroup, numMastered, numGroup)}
          toggleLabel={I18n.t('Toggle outcomes for %{title}', { title: outcomeGroup.title })}
          expanded={expanded}
          onToggle={this.handleToggle}
        >
          <List variant="unstyled" delimiter="solid">
            {
              outcomes.sort(natcompare.byKey('title')).map((outcome) => (
                <ListItem key={outcome.id} margin="0">
                  <Outcome
                    outcome={outcome}
                    expanded={expandedOutcomes.has(outcome.expansionId)}
                    onExpansionChange={onExpansionChange}
                    outcomeProficiency={outcomeProficiency} />
                </ListItem>
              ))
            }
          </List>
        </ToggleGroup>
      </View>
    )
  }
}
