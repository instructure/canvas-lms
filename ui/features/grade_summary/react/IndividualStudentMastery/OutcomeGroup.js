/*
 * Copyright (C) 2018 - present Instructure, Inc.
 *
 * This file is part of Canvas.
 *
 * Canvas is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3 of the License.
 *
 * Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import React from 'react'
import PropTypes from 'prop-types'
import ImmutablePropTypes from 'react-immutable-proptypes'
import I18n from 'i18n!IndividualStudentMasteryOutcomeGroup'
import {View} from '@instructure/ui-view'
import {Flex} from '@instructure/ui-flex'
import {ToggleGroup} from '@instructure/ui-toggle-details'
import {Text} from '@instructure/ui-text'
import {Pill} from '@instructure/ui-pill'
import {List} from '@instructure/ui-list'
import natcompare from '@canvas/util/natcompare'
import TruncateWithTooltip from '../TruncateWithTooltip'
import Outcome from './Outcome'
import * as shapes from './shapes'
import WithBreakpoints, {breakpointsShape} from 'with-breakpoints'

const outcomeGroupHeader = (outcomeGroup, numMastered, numGroup, isVertical) => (
  <Flex
    padding="0 0 0 xxx-small"
    justifyItems={isVertical ? null : 'space-between'}
    direction={isVertical ? 'column' : 'row'}
  >
    <Flex.Item padding="0 x-small 0 0" size={isVertical ? undefined : '0'} grow>
      <Text size="large" weight="bold">
        <TruncateWithTooltip>{outcomeGroup.title}</TruncateWithTooltip>
      </Text>
    </Flex.Item>
    <Flex.Item>
      <Pill text={I18n.t('%{numMastered} of %{numGroup} Mastered', {numMastered, numGroup})} />
    </Flex.Item>
  </Flex>
)

class OutcomeGroup extends React.Component {
  static propTypes = {
    outcomeGroup: shapes.outcomeGroupShape.isRequired,
    outcomes: PropTypes.arrayOf(shapes.outcomeShape).isRequired,
    expanded: PropTypes.bool.isRequired,
    expandedOutcomes: ImmutablePropTypes.set.isRequired,
    onExpansionChange: PropTypes.func.isRequired,
    outcomeProficiency: shapes.outcomeProficiencyShape,
    breakpoints: breakpointsShape
  }

  static defaultProps = {
    outcomeProficiency: null,
    breakpoints: {}
  }

  handleToggle = (_event, expanded) => {
    this.props.onExpansionChange('group', this.props.outcomeGroup.id, expanded)
  }

  render() {
    const {
      outcomeGroup,
      outcomes,
      expanded,
      expandedOutcomes,
      onExpansionChange,
      outcomeProficiency,
      breakpoints
    } = this.props
    const numMastered = outcomes.filter(o => o.mastered).length
    const numGroup = outcomes.length
    const isVertical = !breakpoints.tablet

    return (
      <View as="div" className="outcomeGroup">
        <ToggleGroup
          summary={outcomeGroupHeader(outcomeGroup, numMastered, numGroup, isVertical)}
          toggleLabel={I18n.t('Toggle outcomes for %{title}', {title: outcomeGroup.title})}
          expanded={expanded}
          onToggle={this.handleToggle}
        >
          <List variant="unstyled" delimiter="solid">
            {outcomes.sort(natcompare.byKey('title')).map(outcome => (
              <List.Item key={outcome.id} margin="0">
                <Outcome
                  outcome={outcome}
                  expanded={expandedOutcomes.has(outcome.expansionId)}
                  onExpansionChange={onExpansionChange}
                  outcomeProficiency={outcomeProficiency}
                />
              </List.Item>
            ))}
          </List>
        </ToggleGroup>
      </View>
    )
  }
}

export default WithBreakpoints(OutcomeGroup)
