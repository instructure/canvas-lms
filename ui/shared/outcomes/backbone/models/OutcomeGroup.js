/*
 * Copyright (C) 2023 - present Instructure, Inc.
 *
 * This file is part of Canvas.
 *
 * Canvas is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3 of the License.
 *
 * Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

/* eslint-disable no-void */

import $ from 'jquery'
import {extend} from '@canvas/backbone/utils'
import Backbone from '@canvas/backbone'

import OutcomeCollection from '../collections/OutcomeCollection'

import PaginatedCollection from '@canvas/pagination/backbone/collections/PaginatedCollection'

extend(OutcomeGroup, Backbone.Model)

function OutcomeGroup() {
  return OutcomeGroup.__super__.constructor.apply(this, arguments)
}

OutcomeGroup.prototype.initialize = function (_options) {
  this.setUpOutcomesAndGroupsIfNeeded()
  return OutcomeGroup.__super__.initialize.apply(this, arguments)
}

OutcomeGroup.prototype.name = function () {
  return this.get('title')
}

OutcomeGroup.prototype.isAbbreviated = function () {
  return !this.has('description')
}

OutcomeGroup.prototype.setUrlTo = function (action) {
  return (this.url = function () {
    switch (action) {
      case 'add':
      case 'move':
        return this.get('parent_outcome_group').subgroups_url
      case 'edit':
      case 'delete':
        return this.get('url')
    }
  }.call(this))
}

OutcomeGroup.prototype.setUpOutcomesAndGroupsIfNeeded = function () {
  if (!this.outcomeGroups) {
    this.outcomeGroups = new OutcomeGroupCollection([], {
      parentGroup: this,
    })
  }
  if (!this.outcomes) {
    return (this.outcomes = new OutcomeCollection([]))
  }
}

OutcomeGroup.prototype.getSubtrees = function () {
  return this.outcomeGroups
}

OutcomeGroup.prototype.getItems = function () {
  return this.outcomes
}

OutcomeGroup.prototype.expand = function (force, options) {
  if (force == null) {
    force = false
  }
  if (options == null) {
    options = {}
  }
  this.isExpanded = true
  this.trigger('expanded')
  if (this.expandDfd || force) {
    return $.when()
  }
  this.isExpanding = true
  this.trigger('beginexpanding')
  this.expandDfd = $.Deferred().done(
    (function (_this) {
      return function () {
        _this.isExpanding = false
        return _this.trigger('endexpanding')
      }
    })(this)
  )
  let ref, ref1
  let outcomesDfd
  let outcomeGroupDfd
  if (this.get('outcomeGroups_count') !== 0) {
    outcomeGroupDfd = (ref = this.outcomeGroups) != null ? ref.fetch() : void 0
  }
  if (this.get('outcomes_count') !== 0 && !options.onlyShowSubtrees) {
    outcomesDfd = (ref1 = this.outcomes) != null ? ref1.fetch() : void 0
  }
  return $.when(outcomeGroupDfd, outcomesDfd).done(this.expandDfd.resolve)
}

OutcomeGroup.prototype.collapse = function () {
  this.isExpanded = false
  return this.trigger('collapsed')
}

OutcomeGroup.prototype.toggle = function (options) {
  if (this.isExpanded) {
    return this.collapse()
  } else {
    return this.expand(false, options)
  }
}

extend(OutcomeGroupCollection, PaginatedCollection)

function OutcomeGroupCollection() {
  return OutcomeGroupCollection.__super__.constructor.apply(this, arguments)
}

OutcomeGroupCollection.optionProperty('parentGroup')

OutcomeGroupCollection.prototype.model = OutcomeGroup

OutcomeGroupCollection.prototype.url = function () {
  return this.parentGroup.attributes.subgroups_url
}

export default OutcomeGroup
