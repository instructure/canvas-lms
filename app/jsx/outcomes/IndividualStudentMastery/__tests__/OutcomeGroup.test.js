/*
 * Copyright (C) 2018 - present Instructure, Inc.
 *
 * This file is part of Canvas.
 *
 * Canvas is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3 of the License.
 *
 * Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import React from 'react'
import { render, shallow } from 'enzyme'
import { Set } from 'immutable'
import OutcomeGroup from '../OutcomeGroup'

const outcome = (id, title) => ({
  id,
  title,
  mastered: false,
  mastery_points: 3,
  points_possible: 5,
  ratings: [
    { description: 'My first rating' },
    { description: 'My second rating' }
  ],
  results: [
    {
      id: 1,
      percent: 0.1,
      assignment: {
        id: 2,
        name: 'My alignment',
        html_url: 'http://foo',
        submission_types: ''
      }
    }
  ]
})

const defaultProps = (props = {}) => (
  Object.assign({
    outcomeGroup: {
      id: 10,
      title: 'My group'
    },
    outcomes: [
      {
        id: 1,
        expansionId: 100,
        mastered: false,
        mastery_points: 3,
        points_possible: 5,
        ratings: [
          { description: 'My first rating' },
          { description: 'My second rating' }
        ],
        results: [
          {
            id: 1,
            score: 1,
            percent: 0.1,
            assignment: {
              id: 1,
              html_url: 'http://foo',
              name: 'My assignment',
              submission_types: 'online_quiz'
            }
          }
        ],
        title: 'My outcome'
      }
    ],
    expanded: false,
    expandedOutcomes: Set(),
    onExpansionChange: () => {}
  }, props)
)

it('renders the OutcomeGroup component', () => {
  const wrapper = shallow(<OutcomeGroup {...defaultProps()}/>)
  expect(wrapper.debug()).toMatchSnapshot()
})

describe('header', () => {
  it('includes the outcome group name', () => {
    const wrapper = shallow(<OutcomeGroup {...defaultProps()}/>)
    const header = wrapper.find('ToggleGroup')
    const summary = render(header.prop('summary'))
    expect(summary.text()).toMatch('My group')
  })
})

it('includes the individual outcomes', () => {
  const wrapper = shallow(<OutcomeGroup {...defaultProps()} />)
  expect(wrapper.find('Outcome')).toHaveLength(1)
})

it('renders correctly expanded', () => {
  const wrapper = shallow(<OutcomeGroup {...defaultProps()} expanded />)
  expect(wrapper.debug()).toMatchSnapshot()
})

it('renders outcomes in alphabetical order by title', () => {
  const props = defaultProps({
    outcomes: [
      outcome(1, 'ZZ Top'),
      outcome(2, 'Aardvark'),
      outcome(3, 'abba'),
      outcome(4, 'Aerosmith')
    ]
  })
  const wrapper = shallow(<OutcomeGroup {...props} />)
  const outcomes = wrapper.find('Outcome')
  expect(outcomes).toHaveLength(4)
  expect(outcomes.get(0).props.outcome.title).toEqual('Aardvark')
  expect(outcomes.get(1).props.outcome.title).toEqual('abba')
  expect(outcomes.get(2).props.outcome.title).toEqual('Aerosmith')
  expect(outcomes.get(3).props.outcome.title).toEqual('ZZ Top')
})

describe('handleToggle()', () => {
  it('calls the correct onExpansionChange callback', () => {
    const props = defaultProps()
    props.onExpansionChange = jest.fn()
    const wrapper = shallow(<OutcomeGroup {...props} />)
    wrapper.instance().handleToggle(null, true)
    expect(props.onExpansionChange).toBeCalledWith('group', 10, true)
  })
})
