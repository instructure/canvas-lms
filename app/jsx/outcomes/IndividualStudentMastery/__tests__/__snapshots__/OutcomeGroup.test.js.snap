// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly expanded 1`] = `
"<View as=\\"div\\" className=\\"outcomeGroup\\" display=\\"auto\\">
  <ToggleGroup summary={{...}} toggleLabel=\\"Toggle outcomes for My group\\" expanded={true} onToggle={[Function]} size=\\"medium\\" icon={[Function]} iconExpanded={[Function]} defaultExpanded={false} transition={true} as=\\"span\\" elementRef={[Function]} border={true}>
    <List variant=\\"unstyled\\" delimiter=\\"solid\\" as=\\"ul\\" margin=\\"none\\" size=\\"medium\\">
      <ListItem margin=\\"0\\">
        <Outcome outcome={{...}} expanded={false} onExpansionChange={[Function]} outcomeProficiency={{...}} />
      </ListItem>
    </List>
  </ToggleGroup>
</View>"
`;

exports[`renders the OutcomeGroup component 1`] = `
"<View as=\\"div\\" className=\\"outcomeGroup\\" display=\\"auto\\">
  <ToggleGroup summary={{...}} toggleLabel=\\"Toggle outcomes for My group\\" expanded={false} onToggle={[Function]} size=\\"medium\\" icon={[Function]} iconExpanded={[Function]} defaultExpanded={false} transition={true} as=\\"span\\" elementRef={[Function]} border={true}>
    <List variant=\\"unstyled\\" delimiter=\\"solid\\" as=\\"ul\\" margin=\\"none\\" size=\\"medium\\">
      <ListItem margin=\\"0\\">
        <Outcome outcome={{...}} expanded={false} onExpansionChange={[Function]} outcomeProficiency={{...}} />
      </ListItem>
    </List>
  </ToggleGroup>
</View>"
`;
